File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.java
class Solution {
    public long findKthSmallest(int[] coins, int k) {
        
        long minCoin = Long.MAX_VALUE;

        for (int coin : coins)
            minCoin = Math.min(minCoin, coin);
        
        long left = 1;
        long right = minCoin * k;

        while (left < right) {
            long mid = left + (right - left) / 2;

            if (count(mid, coins, 0, 1, 0) >= k)
                right = mid;
            else
                left = mid + 1;
            
        }

        return left;
    }

    // helper function for count
    private long count(long x, int[] coins, int index, long lcm, int selected) {
        long result = 0;

        for (int i = index; i < coins.length; i++) {

            long newLcm = lcm(lcm, coins[i]);

            if (newLcm > x)
                continue;
            
            long contribution = x / newLcm;

            if ((selected + 1) % 2 == 1)
                result += contribution;
            
            else
                result -= contribution;
            
            result += count(x, coins, i + 1, newLcm, selected + 1);
        }

        return result;
    }

    // helper function for gcd
    private long gcd(long a, long b) {
        while (b != 0) {
            long temp = a % b;
            a = b;
            b = temp;
        }
        return a;
    }

    // helper function for lcm
    private long lcm(long a, long b) {
        return a / gcd(a, b) * b;
    }
}

File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.py
class Solution:
    def findKthSmallest(self, coins: list[int], k: int) -> int:
        minCoin = min(coins)
        
        left = 1
        right = minCoin * k

        def gcd(a: int, b: int) -> int:
            while b != 0:
                a, b = b, a % b
            return a

        def lcm(a: int, b: int) -> int:
            return (a // gcd(a, b)) * b

        def count(x: int, index: int, current_lcm: int, selected: int) -> int:
            result = 0
            for i in range(index, len(coins)):
                new_lcm = lcm(current_lcm, coins[i])
                
                if new_lcm > x:
                    continue
                
                contribution = x // new_lcm
                
                if (selected + 1) % 2 == 1:
                    result += contribution
                else:
                    result -= contribution
                    
                result += count(x, i + 1, new_lcm, selected + 1)
                
            return result

        while left < right:
            mid = left + (right - left) // 2
            
            if count(mid, 0, 1, 0) >= k:
                right = mid
            else:
                left = mid + 1
                
        return left
